Fix creation performance of form element menuparent, (slow down in menu item edit form, for item that belongs to large menu)

ggppdk opened the pull request:
Fix creation performance of **form element menuparent**
- used in menu item edit form
### Summary of Changes

Remove unused left join in getOptions for menuparent form element

No reason because for it being there because
1. Join is left
2. The results of the join are not used anywhere neither in select nor where and not in any other clause
3. There is no PHP code that adds some usage of it under some condition
### Testing Instructions

Visit menu item edit form and, check parent menuitem form field is as before
### Documentation Changes Required

ggppdk committed:
Remove unused left join in getOptions for menuparent form element

ggppdk committed:
Replaced unneeded group by with distinct

ggppdk commented:
1 more optimization, replaced unneeded
**group by with distinct**

For reasoning behind the above, 
- and for when the above can be done without problems

please see my comment here for similar fix in :
https://github.com/joomla/joomla-cms/issues/9516#issuecomment-200289018

which was followed by @alikon making a similar PR here , that was merged:
https://github.com/joomla/joomla-cms/pull/9689

brianteeman commented:
Needs to be checked on all databases

On 16 August 2016 at 12:05, Georgios Papadakis <notifications@github.com>
wrote:

> 1 more optimization, replaced unneeded
> *group by with distinct*
>
> For reasoning behind the above,
>
>    - and for when the above can be done without problems
>
> please see my comment here for similar fix in :
> #9516 (comment)
> <https://github.com/joomla/joomla-cms/issues/9516#issuecomment-200289018>
>
> which was followed by @alikon <https://github.com/alikon> making a
> similar PR here , that was merged:
> #9689 <https://github.com/joomla/joomla-cms/pull/9689>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11628#issuecomment-240071901>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPH8TOWzIewyv1IKoTNKMeoDiJfOOsIks5qgZlngaJpZM4JlNVu>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

jeckodevelopment commented:
> Needs to be checked on all databases

@alikon can you help with this, please?

prathumwan commented:
could not measure it with debug because running out of memory, but if you see the result done by hand it's a big improvement.

select order by result:
![joomla patch 11628 before](https://cloud.githubusercontent.com/assets/16284417/17746451/47f0640a-64da-11e6-8b31-c45bfafb901a.png)

select distinct result:
![joomla patch 11628 after](https://cloud.githubusercontent.com/assets/16284417/17746501/7ccead9e-64da-11e6-818a-bebda2c34a35.png)

alikon commented:
in postgresql don't work as it is now , it give an **SQL ERROR**
 for the `ORDER BY` clause field is used `a.lft` and therefore must be in the select list of fields when the `SELECT` use `DISTINCT()`

@ggppdk i've submitted a PR https://github.com/ggppdk/joomla-cms/pull/2 to your branch with the fix for postgresql

@jeckodevelopment always happy to help when free time permits ;)

alikon committed:
fix #11628 for postgresql (#2)

in postgresql don't work as it is now , it give an **SQL ERROR**
 for the `ORDER BY` clause field used `a.lft` must be in the select list of fields when the `SELECT` use `DISTINCT()`

ggppdk commented:
@alikon 
Merged

alikon commented:
@ggppdk thanks, 
for the perfomance, have someone, some "decent" menu table dump to meter ?

p.s.
this query is still ugly  ` a.published != -2` for example. etc...

jeckodevelopment commented:
@alikon you're our DB expert :) thank you

ggppdk commented:
> this query is still ugly a.published != -2 for example. etc...

ok but nothing significant to further improve performance

**these 2 changes are enough to fix the performance issue**, on large menu (with a few thousands of menu items)
- removal of unsed self-join with the full table (**unlike to the parent join that limits to -specific- row id**)
- the replacement of group-by with distinct , since we do not do use any aggregating function to calculate something

alikon commented:
I have tested this item :white_check_mark: successfully on 2972851a1c8f95436f57294722ed2790b8f3e547<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11628">issues.joomla.org/joomla-cms/11628</a>.</sub>

prathumwan commented:
I have tested this item :white_check_mark: successfully on 2972851a1c8f95436f57294722ed2790b8f3e547<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11628">issues.joomla.org/joomla-cms/11628</a>.</sub>

andrepereiradasilva commented:
can we have RTC here?

brianteeman commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11628">issues.joomla.org/joomla-cms/11628</a>.</sub>
